federation/pkg/federation-controller/job/jobcontroller.go

```diff
@@ -0,0 +1,538 @@
+					if fedStatusCompleteCondition == nil ||
+						fedStatusCompleteCondition.LastTransitionTime.Before(condition.LastTransitionTime) {
+						fedStatusCompleteCondition = &condition
+					}
+				} else if condition.Type == batchv1.JobFailed {
+					if fedStatusFailedCondition == nil ||
+						fedStatusFailedCondition.LastTransitionTime.Before(condition.LastTransitionTime) {
+						fedStatusFailedCondition = &condition
+					}
+				}
+			}
+			if currentLjob.Status.StartTime != nil {
+				if fedStatus.StartTime == nil || fedStatus.StartTime.After(currentLjob.Status.StartTime.Time) {
+					fedStatus.StartTime = currentLjob.Status.StartTime
+				}
+			}
+			if currentLjob.Status.CompletionTime != nil {
+				if fedStatus.CompletionTime == nil || fedStatus.CompletionTime.Before(*currentLjob.Status.CompletionTime) {
+					fedStatus.CompletionTime = currentLjob.Status.CompletionTime
+				}
+			}
+			fedStatus.Active += currentLjob.Status.Active
+			fedStatus.Succeeded += currentLjob.Status.Succeeded
+			fedStatus.Failed += currentLjob.Status.Failed
+		}
+	}
+
+	// federated job fails if any local job failes
+	if fedStatusFailedCondition != nil {
+		fedStatus.Conditions = append(fedStatus.Conditions, *fedStatusFailedCondition)
+	} else if fedStatusCompleteCondition != nil {
+		fedStatus.Conditions = append(fedStatus.Conditions, *fedStatusCompleteCondition)
+	}
+	if !reflect.DeepEqual(fedStatus, fjob.Status) {
+		fjob.Status = fedStatus
+		_, err = fjc.fedClient.Batch().Jobs(fjob.Namespace).UpdateStatus(fjob)
+		if err != nil {
+			return statusError, err
+		}
+	}
+
+	if len(operations) == 0 {
+		// Everything is in order
+		return statusAllOk, nil
+	}
+
+	if glog.V(4) {
+		for i, op := range operations {
+			job := op.Obj.(*batchv1.Job)
+			glog.V(4).Infof("operation[%d]: %s, %s/%s/%s, %d", i, op.Type, op.ClusterName, job.Namespace, job.Name, *job.Spec.Parallelism)
+		}
+	}
+	err = fjc.fedUpdater.UpdateWithOnError(operations, updateTimeout, func(op fedutil.FederatedOperation, operror error) {
+		fjc.eventRecorder.Eventf(fjob, api.EventTypeNormal, "FailedUpdateInCluster", "Job update in cluster %s failed: %v", op.ClusterName, operror)
+		glog.Errorf("Failed to execute updates for %s %s/%s: %v", op.Type, op.ClusterName, key, operror)
+	})
+	if err != nil {
+		return statusError, err
+	}
+
+	// Some operations were made, reconcile after a while.
+	return statusNeedRecheck, nil
+
+}
+
+func (fjc *JobController) reconcileJobsOnClusterChange() {
+	if !fjc.isSynced() {
+		fjc.clusterDeliverer.DeliverAfter(allClustersKey, nil, clusterAvailableDelay)
+	}
+	jobs := fjc.jobStore.List()
+	for _, rs := range jobs {
+		key, _ := controller.KeyFunc(rs)
+		glog.Infof("deliverCluster: ")
+		fjc.deliverJobByKey(key, 0, false)
+	}
+}
+
+func (frsc *JobController) deleteJob(key string) error {
+	glog.Infof("deleting job: %v", key)
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	// try delete from all clusters
```

Comment:
Use deletion helper for this. We will delete from underlying clusters only if DeleteOptions.OrphanDependents = false